### PR TITLE
8266798: C1: More types of instruction can also apply LoopInvariantCodeMotion

### PR DESCRIPTION
**Pull request by kelthuzadx**
C1 only applies LoopInvariantCodeMotion for instructions whose types are Constant/ArithmeticOp/LoadField/ArrayLength/LoadIndexed. We are possible to apply this optimization for more types of instruction.

Candidates: NegateOp,Convert.

Due to the lack of verification at IR level, it is difficult to write jtreg to check if it transformed, so I can only demonstrate it with a simple program:
```java
// run with -XX:+PrintValueNumbering
static int foo10(int t){
    int sum=12;
    for(int i=0;i<100;i++){
        sum += 12;
        sum += -t;
        sum += (long)t;
    }
    return sum;
}
```
Before:
```
[...]
* loop invariant code motion for short loop B1
processing block B1
Value Numbering: insert Constant i10  (size 11, entries 3, nesting 2)
Instruction i10 is loop invariant
processing block B2
Value Numbering: Constant i12 equal to i5  (size 11, entries 3, nesting-diff 1)
substitution for 12 set to 5
Instruction i12 is loop invariant   // only 12 is recongized
Value Numbering: insert Constant i20  (size 11, entries 4, nesting 2)
Instruction i20 is loop invariant
** loop successfully optimized
[...]
```
After:
```
[...]
** loop invariant code motion for short loop B1
processing block B1
Value Numbering: insert Constant i10  (size 11, entries 3, nesting 2)
Instruction i10 is loop invariant
  6    0    i10    100
processing block B2
Value Numbering: Constant i12 equal to i5  (size 11, entries 3, nesting-diff 1)
substitution for i12 set to 105
Instruction i12 is loop invariant
  11   0    i12    12
Instruction i14 is loop invariant
. 16   0    i14    -i4
Value Numbering: insert Convert l17  (size 11, entries 4, nesting 2)
Instruction l17 is loop invariant
. 22   0    l17    i2l(i4)
Value Numbering: insert Constant i20  (size 11, entries 5, nesting 2)
Instruction i20 is loop invariant
  26   0    i20    1
[...]
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266798](https://bugs.openjdk.java.net/browse/JDK-8266798): C1: More types of instruction can also apply LoopInvariantCodeMotion

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3965/head:pull/3965` \
`$ git checkout pull/3965`

Update a local copy of the PR: \
`$ git checkout pull/3965` \
`$ git pull https://git.openjdk.java.net/jdk pull/3965/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3965`

View PR using the GUI difftool: \
`$ git pr show -t 3965`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3965.diff">https://git.openjdk.java.net/jdk/pull/3965.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/c1/c1_ValueMap.cpp
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_ValueMap.cpp:592]: Shouldn't the third argument be a char and the fourth an int?

**Review by TobiHartmann — CHANGES_REQUESTED** [review]

File: src/hotspot/share/c1/c1_ValueMap.cpp
Comment (kelthuzadx) [reply src/hotspot/share/c1/c1_ValueMap.cpp:592]: Good catch, fixed.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by kelthuzadx** [comment]
Thank you Tobias for the review!

/integrate

**Comment by TobiHartmann** [comment]
This is non-trivial and requires a second review.

**Review by neliasso — APPROVED** [review]
Looks good.

**Comment by kelthuzadx** [comment]
Thank you Tobias and Nils!

**Pull request closed without merging**
